### ISSUE TITLE
Add option to pass IDPList as authentication parameter

### ISSUE THREAD

**Issue by bartlangelaan** [simplesamlphp/simplesamlphp#1562]
**Is your feature request related to a problem? Please describe.**
We are currently implementing with an IDP that _sometimes_ requires scoping IDPs.

**Describe the solution you'd like**
Allow the property `IDPList` to be provided to `$auth->login()`.

**Describe alternatives you've considered**
Adding the `saml:IDPList` property (undocumented) does work _if_ this statement is adjusted to it always returns false: https://github.com/simplesamlphp/simplesamlphp/blob/959aa29de0830d3f792f6b1f785158f08e67e0a9/modules/saml/lib/Auth/Source/SP.php#L731

But I don't think using `saml:IDPList` is the right option, is it?

**Comment by bartlangelaan** [comment]
No, we are not. We are using an [HM](https://afsprakenstelsel.etoegang.nl/display/as/Interface+specifications+DV-HM), that implements various other IDPs.

We want to skip the HM screen where you can select the IDP you want to use. That is possible if the Scoping element is used, with only one IDPEntry.

**Comment by tvdijen** [comment]
Then all you gotta do is set the `IDPList` in your SP authsource and you're set?

**Comment by bartlangelaan** [comment]
But, I only want to do it _sometimes_.

In some cases, the IDP is preset, so we want to skip the HM screen.

In other cases, the IDP is not predefined, so we want to show the HM screen.

We would like to use the same SP for that. So that's why we want to provide it in the `$sp->login()` method, instead of statically in the SP authsource.

**Comment by tvdijen** [comment]
Ah, I see... But that should already work if you pass it to `$sp->login()` as `saml:IDPList`...
In fact you're now adding `$state['IDPList']`, while the exact same content should already be in the `$IDPList`-array (added in line 534)...

So, the suggested fix seems wrong and we have to figure out what's going on around line 731 in the `authenticate()` method.  The phpdoc kinda already reveals that it assumes we're a proxy because we have an IDPList set, which is a false assumption.. I'll try and investigate this soon.

**Comment by thijskh** [comment]
We had this issue at hand before so I've looked into it before. Have to dig a bit for the specifics though.

There are two scenarios that work with IDPList:

**The proxy setup.** That is, SSP is a proxy, it receives an incoming SAML authnrequest with an IDPList, that should be correctly construed as an attempt to filter the available IdPs SSP knows about and select which to use for its outoing SP function. The saml:IDPlist state element is populated when receiving an AuthnRequest in the IdP role. The SP code will select either 1 IdP or forward to the disco with the (filtered) list to choose from. So I think [this code](https://github.com/simplesamlphp/simplesamlphp/blob/959aa29de0830d3f792f6b1f785158f08e67e0a9/modules/saml/lib/Auth/Source/SP.php#L731) is not wrong per se and also not "undocumented", as it just implements a published SAML spec aspect.

**A regular (end) SP setup** that wants to send an IDPList in the outgoing authentication request. These do not need to be IdPs this SP knows about indeed, it just wants to add an element to the authnrequest. This is currently possible in the entity configuration. I think [this code](https://github.com/simplesamlphp/simplesamlphp/blob/959aa29de0830d3f792f6b1f785158f08e67e0a9/modules/saml/lib/Auth/Source/SP.php#L536) is indeed wrong. If any saml:IDPList was received from a remote SP, we do not need to pass it on to any downstream IdP. This use actually might be intended to allow runtime configuration, but because it overloads the already in use name "saml:IDPList" that was used for externally sourced info, this cannot work. Because it's not documented, I'm not sure if it was actually intended at all that you could set this at runtime. But it seems a reasonable explanation.

The code is in fact ancient and stems more or less unmodified from SSP 1.6.

It seems logical to me not to overload `saml:IDPList` state element but to keep the two use cases clearly separate, since the information comes from a different source _and_ has a different goal.
So the approach to add a new runtime setting to configure the IDPList does not strike me as a bad idea.

**Comment by Paragin** [comment]
We are currently having trouble with implementing the exact same solution as @bartlangelaan - in our case we need to use scoping to be able to skip a WAYF-screen and direct students directly to the login screen of their school. Given that it it depends on the origin of the user which IDP we need to add to the scope, it's not possible for us to add the IDPList to the authsource.

Is there anything we can do to get this issue sorted?

**Comment by Paragin** [comment]
I can do that.. :) 

Since I'm not 100% familiar with the code: [at the lines you mentiond](https://github.com/simplesamlphp/simplesamlphp/blob/959aa29de0830d3f792f6b1f785158f08e67e0a9/modules/saml/lib/Auth/Source/SP.php#L536), there is also a reference to saml:ProxyCount. I would assume that the `saml:` prefix can be removed there as well to keep everything consistent with the [documentation about scoping](https://simplesamlphp.org/docs/1.16/simplesamlphp-scoping)?

**Comment by thijskh** [comment]
No, the ProxyCount is indeed correctly related and should be carried over when operating as a proxy. I would focus on changing the IDPList code as per the analysis earlier in this issue.

**Comment by Paragin** [comment]
I've created a new PR for this: https://github.com/simplesamlphp/simplesamlphp/pull/1653
It would be really nice to get this into 1.19 given that 2.0 might take a while. Once this is approved, shall I create a PR for 1.19 as well?
